File: sources/pascal/06_char_count.pas
{ task 06 char_count — expected output: 10000000 }
{ build: fpc -O3 -oprogram 06_char_count.pas    run: ./program }
{ Windows x64: i386-win32 fpc + cross.x86_64-win64 add-on, build with -Px86_64; run as program.exe }

program prog;
{$mode objfpc}{$H+}

uses
  SysUtils;

const
  Pattern: array[0..9] of AnsiChar = 'abcdefghij';

var
  text: AnsiString;
  i, count: Int64;
begin
  { one hundred million characters, built in one go: the ten character block is
    moved into every tenth slot, so the build is not part of the measurement }
  SetLength(text, 100000000);
  for i := 0 to 9999999 do
    Move(Pattern[0], text[i * 10 + 1], 10);

  count := 0;
  for i := 1 to 100000000 do
    if text[i] = 'h' then
      Inc(count);
  WriteLn(count);
end.

File: sources/swift/06_char_count.swift
// task 06 char_count — expected output: 10000000
// build: swiftc -O -o prog 06_char_count.swift    run: ./prog

// The 100 MB text is built once, by repeating the ten-character block.
let text = String(repeating: "abcdefghij", count: 10_000_000)

// The scan walks the UTF-8 bytes of the string directly; materialising the
// characters into an Array would allocate a hundred million elements first.
var count = 0
text.withCString { ptr in
    for i in 0..<100_000_000 {
        let ch = ptr[i]
        if ch == 0x61 {          // 'a' — skip
            continue
        } else if ch == 0x65 {   // 'e' — skip
            continue
        } else if ch == 0x68 {   // 'h' — count
            count += 1
        }
    }
}

print(count)
